File: go/tree/94_inorder_traversal.go
package tree

// inorderTraversal 中序遍历 （左根右），递归实现
func inorderTraversal(root *TreeNode) []int {

	ans := make([]int, 0)
	if root == nil {
		return ans
	}

	var inorder func(*TreeNode)

	inorder = func(root *TreeNode) {
		if root == nil {
			return
		}

		inorder(root.Left)
		ans = append(ans, root.Val)
		inorder(root.Right)
	}

	inorder(root)
	return ans
}

// inorderTraversal_ 中序遍历的迭代实现
func inorderTraversal_(root *TreeNode) []int {

	ans := make([]int, 0)

	stack := make([]*TreeNode, 0)

	node := root

	for len(stack) > 0 || node != nil {

		// 左根右顺序，所以要一直找左孩子
		// 先访问左子树
		for node != nil {
			stack = append(stack, node)
			node = node.Left
		}
		// 走出上面循环，node == nil， 左子树放完了，访问改节点，然后右子树
		// 最后一个左孩子即stack 中的最后一个元素
		popped := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		ans = append(ans, popped.Val)
		node = popped.Right
	}

	return ans
}

File: rust/src/tree/_94_inorder_traversal.rs
use super::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

// 迭代实现
pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut ans = vec![];
    let mut stack = vec![];

    let mut node = root;

    while !stack.is_empty() || node.is_some() {
        // 中序遍历，将左孩子压栈
        while let Some(n) = node {
            // 取出左孩子
            node = n.borrow_mut().left.take();
            // 把当前节点压栈，后续出栈 取右孩子
            stack.push(n);
        }

        // 当node 为None时，意味着stack 中最后一个元素为最左的孩子
        if let Some(n) = stack.pop() {
            ans.push(n.borrow().val);
            node = n.borrow_mut().right.take();
        }
    }

    ans
}

// inorder_traversal 递归实现
pub fn inorder_traversal2(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let ans = inorder(&root);

    ans
}

fn inorder(root: &Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    if let Some(node) = root {
        let mut left = inorder(&node.borrow().left);

        let mut r = vec![node.borrow().val];

        let mut right = inorder(&node.borrow().right);

        left.append(&mut r);
        left.append(&mut right);
        left
    } else {
        vec![]
    }
}
